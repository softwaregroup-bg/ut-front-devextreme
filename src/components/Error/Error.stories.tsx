import React from 'react';
import { withReadme } from 'storybook-readme';

// @ts-ignore: md file and not a module
import README from './README.md';
import Error from './index';

export default {
    title: 'Error',
    component: Error,
    decorators: [withReadme(README)],
    args: {
        state: {
            error: {
                open: true
            }
        }
    }
};

export const Basic = () => <Error message='Error message' />;
export const InvalidCredentials = () => <Error />;
export const InterpolateProps = () => <Error message='Error {message}' params={{message: 'message'}} />;
export const InterpolateRedux = () => <Error />;

InvalidCredentials.args = {
    state: {
        error: {
            open: true,
            message: 'Invalid credentials',
            type: 'identity.invalidCredentials'
        }
    }
};

InterpolateRedux.args = {
    state: {
        error: {
            open: true,
            message: 'Error {message}',
            type: 'core.throttle',
            params: {
                message: 'dynamic'
            },
            title: 'Error'
        }
    }
};
