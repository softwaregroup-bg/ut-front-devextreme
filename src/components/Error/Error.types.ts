import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';
import error from '../images/error.png';

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
    message?: React.ReactNode,
    params?: {}
}

const styles = createStyles({
    errorIconWrap: {
        display: 'block',
        width: 40,
        margin: '0 auto',
        textAlign: 'center',
        marginBottom: 10
    },
    errorIcon: {
        width: 40,
        height: 40,
        background: `url(${error})`
    },
    errorMessageWrap: {
        display: 'block',
        width: '100%',
        textAlign: 'center',
        color: '#E84949'
    },
    errorButtonWrap: {
        display: 'block',
        margin: 10,
        textAlign: 'center'
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
