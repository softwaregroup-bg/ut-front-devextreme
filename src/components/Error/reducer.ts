import {LOGOUT} from '../Login/actions';

import errorMessage from './errorMessage';

const defaultState = {open: false, title: '', message: '', statusCode: 200, type: '', params: {}};

export default (state = defaultState, action) => {
    if (['front.error.close', LOGOUT].includes(action.type)) {
        return {...state, ...defaultState};
    }
    if (action.error) {
        if (action.suppressErrorWindow) return state;
        const msg = (errorMessage(action.error) || errorMessage(state.message));
        const params = action.error.params || {};
        const statusMessage = action.error.statusMessage;
        const statusCode = action.error.statusCode;
        let title = 'Error';
        const type = action.error.type;
        if (statusMessage) {
            title = `${statusMessage}`;
            if (statusCode) {
                title += `(${statusCode})`;
            }
        }

        return {
            ...state,
            open: true,
            title: title,
            message: msg,
            statusCode,
            type,
            params
        };
    }
    return state;
};
