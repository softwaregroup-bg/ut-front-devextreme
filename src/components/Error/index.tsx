import React from 'react';
import clsx from 'clsx';
import { useSelector, useDispatch } from 'react-redux';
import interpolate from 'ut-function.interpolate';
import { Dialog, Button } from '../prime';

import Text from '../Text';
import { logout } from '../Login/actions';

import { Styled, StyledType } from './Error.types';
import { State } from '../Store/Store.types';

const selectError = (state: State) => state.error;
const breakpoints = {'960px': '75vw', '640px': '95vw'};
const width = {width: '30vw'};

const Error: StyledType = ({ classes, message: errorMessage, params: errorParams }) => {
    let closable = true;
    let actionButtons;

    const {open, title: header, message, type, statusCode, params} = useSelector(selectError);
    const dispatch = useDispatch();
    const unauthorized = (type === 'identity.invalidCredentials') || (statusCode === 401);
    const handleClose = React.useCallback(() => dispatch(unauthorized ? logout() : {type: 'front.error.close'}), [dispatch, unauthorized]);

    if (unauthorized) {
        closable = false;
        actionButtons = <Button label="Login" onClick={handleClose} />;
    } else {
        actionButtons = <Button label="Close" onClick={handleClose} />;
    }

    return open ? (
        <Dialog
            visible={open}
            onHide={handleClose}
            breakpoints={breakpoints}
            style={width}
            {...{closable, header}}
        >
            <div className={clsx(classes.errorIconWrap, classes.errorIcon)} />
            <div className={classes.errorMessageWrap}>
                <Text params={errorParams || params} interpolate={interpolate}>{errorMessage || message}</Text>
            </div>
            <div className={classes.errorButtonWrap}>
                {actionButtons}
            </div>
        </Dialog>
    ) : null;
};

export default Styled(Error);
