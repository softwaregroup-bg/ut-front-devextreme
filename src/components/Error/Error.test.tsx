import React from 'react';
import {waitFor} from '@testing-library/react';

import { render } from '../test';
import Error from './index';

describe('<Error />', () => {
    it('render equals snapshot', async() => {
        const { getByRole } = render(<Error />, {error: {open: true}});
        await waitFor(() => expect(getByRole('dialog').querySelector('div.p-dialog-title')).toBeTruthy());
        expect(getByRole('dialog')).toMatchSnapshot();
    });
    it('interpolate equals snapshot', async() => {
        const { getByText } = render(<Error />, {error: {open: true, message: 'Error {message}', params: {message: 'dynamic param'}}});
        await waitFor(() => expect(getByText('Error dynamic param')).toBeTruthy());
        expect(getByText('Error dynamic param')).toMatchSnapshot();
    });
});
