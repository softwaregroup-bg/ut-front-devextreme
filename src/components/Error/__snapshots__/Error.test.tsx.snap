// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Error /> interpolate equals snapshot 1`] = `
<span>
  Error dynamic param
</span>
`;

exports[`<Error /> render equals snapshot 1`] = `
<div
  aria-describedby="pr_id_2_content"
  aria-labelledby="pr_id_2_header"
  aria-modal="true"
  class="p-dialog p-component p-dialog-enter p-dialog-enter-active"
  id="pr_id_2"
  pr_id_1=""
  role="dialog"
  style="width: 30vw;"
>
  <div
    class="p-dialog-header"
  >
    <div
      class="p-dialog-title"
      id="pr_id_2_header"
    />
    <div
      class="p-dialog-header-icons"
    >
      <button
        aria-label="Close"
        class="p-dialog-header-icon p-dialog-header-close p-link"
        type="button"
      >
        <span
          class="p-dialog-header-close-icon pi pi-times"
        />
      </button>
    </div>
  </div>
  <div
    class="p-dialog-content"
    id="pr_id_2_content"
  >
    <div
      class="Error-errorIconWrap-1 Error-errorIcon-2"
    />
    <div
      class="Error-errorMessageWrap-3"
    >
      <span />
    </div>
    <div
      class="Error-errorButtonWrap-4"
    >
      <button
        class="p-button p-component"
      >
        <span
          class="p-button-label p-c"
        >
          Close
        </span>
      </button>
    </div>
  </div>
  <div
    class="p-resizable-handle"
    style="z-index: 90;"
  />
</div>
`;
